payment-account-age-witness.adoc

```diff
@@ -0,0 +1,196 @@
+
+
+=== AgeWitness data object
+
+The AgeWitness object contains a hash, a signature, the pubKey and the tradeDate. The hash gets created with Sha256. Input is the ageWitnessInputData concatenated with a 256 bytes salt. The ageWitnessInputData is the smallest set of uniquely identifying payment account data (e.g. concatenated IBAN and BIC). We don't use the complete payment account data because we don't want to break an existing ageWitness by minor changes like changing the holder name (e.g. adding middle name).
+
+We add also a signature of the hash and the public key for assuring that the AgeWitness data cannot be used by anyone else (see: <<hijacking>>). The application wide 1024 bit DSA signing key is used. Signature algorith is "SHA256withDSA".
+
+The tradeDate must not be older than 1 day compared to receiving peer's local date. If the data is too old the AgeWitness object will get rejected. With that we protect against back-dating attempts (see: <<back-dating>>). We allow a rather large tolerance because computer clocks might be out of sync and the relevant periods are rather long (1 month, 2 months), so the max. gain from an abuse of that tolerance window of 1 day is negligible.
+
+----
+    // class PaymentAccountAgeWitnessService
+    public PaymentAccountAgeWitness getPaymentAccountWitness(PaymentAccount paymentAccount, Trade trade) throws CryptoException {
+        byte[] ageWitnessInputData = paymentAccount.getAgeWitnessInputData();
+        byte[] salt = paymentAccount.getSalt();
+        final byte[] combined = ArrayUtils.addAll(ageWitnessInputData, salt);
+        byte[] hash = Sha256Hash.hash(combined);
+        byte[] signature = Sig.sign(keyRing.getSignatureKeyPair().getPrivate(), hash);
+        long tradeDate = trade.getTakeOfferDate().getTime();
+        return new PaymentAccountAgeWitness(hash,
+                keyRing.getPubKeyRing().getSignaturePubKeyBytes(),
+                signature,
+                tradeDate);
+    }
+
+    // getAgeWitnessInputData at example class SepaAccount
+    public byte[] getAgeWitnessInputData() {
+        return ArrayUtils.addAll(getIban().getBytes(), getBic().getBytes());
+    }
+
+    // Util for getting salt
+    public static byte[] getSalt(int size) {
+        byte[] salt = new byte[size];
+        new SecureRandom().nextBytes(salt);
+        return salt;
+    }
+
+    // VO class PaymentAccountAgeWitness contains:
+    private final byte[] hash;
+    private final byte[] pubKeyBytes;
+    private final byte[] signaturePubKeyBytes;
+    private final long tradeDate;
+----
+
+
+=== AgeWitness propagation
+
+The salt value will be locally persisted with the payment account object. The trader will *only publish the AgeWitness object at the first trade*!
```

Comment:
another random thought - in this design there is only one salt per user.
Wouldn't it be possible to derive the salt from his seed words? Suppose he reinstalls bisq and loses the salt, we can recreate it by taking a hash of his private key + account data? Just a thought and one less thing to store.

Reply:
Good point. But I think we should even make it accessible for backup/export, so if the user starts over again with a new app /wallet he can keep his ageWitness. Changing the wallet increases privacy (bloom filter) so binding it to the seed words might be not so good. But exporting the payment account is already supported and there we can add the salt. But it need to be set-able as well if the user re-creates an account with same iban/bic he wants to use the same salt. So that need to be readable and set-able at account creation.
Will add that to the doc.

Reply:
Added:

=== Salt management

If the user changes his payment account or start over with a new application we need to support that he can re-use the salt he used with a certain bank account. One option would be to add an extra field in the payment account setup screen where the user can add a past salt (otherwise the app generates a random salt).